x/protocolpool/keeper/keeper.go

```diff
@@ -67,3 +88,115 @@ func (k Keeper) GetCommunityPool(ctx context.Context) (sdk.Coins, error) {
 	}
 	return k.bankKeeper.GetAllBalances(ctx, moduleAccount.GetAddress()), nil
 }
+
+func (k Keeper) claimFunds(ctx context.Context, recipient sdk.AccAddress) (amount sdk.Coin, err error) {
+	// get claimable funds from distribution info
+	amount, err = k.getClaimableFunds(ctx, recipient)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	// distribute amount from feepool
+	err = k.DistributeFromFeePool(ctx, sdk.NewCoins(amount), recipient)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	return amount, nil
+}
+
+func (k Keeper) getClaimableFunds(ctx context.Context, recipient sdk.AccAddress) (amount sdk.Coin, err error) {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+
+	budget, err := k.BudgetProposal.Get(ctx, recipient)
+	if err != nil {
+		if errors.Is(err, collections.ErrNotFound) {
+			return sdk.Coin{}, fmt.Errorf("no budget found for recipient: %s", recipient.String())
+		}
+		return sdk.Coin{}, err
+	}
+
+	// check if the distribution is completed
+	if budget.Tranches <= 0 {
+		// remove the entry of budget ended recipient
+		err := k.BudgetProposal.Remove(ctx, recipient)
+		if err != nil {
+			return sdk.Coin{}, err
+		}
+		// Return the end of the budget
+		return sdk.Coin{}, fmt.Errorf("budget ended for recipient: %s", recipient.String())
+	}
+
+	currentTime := sdkCtx.BlockTime().Unix()
+	startTime := budget.StartTime
+	// Check if the start time is reached
+	if uint64(currentTime) < startTime {
+		return sdk.Coin{}, fmt.Errorf("distribution has not started yet")
+	}
+
+	if budget.LastClaimTime == 0 {
+		budget.LastClaimTime = budget.StartTime
+	}
+
+	// Calculate the time elapsed since the last claim time
+	timeElapsed := uint64(currentTime) - budget.LastClaimTime
+
+	// Check the time elapsed has passed period length
+	if timeElapsed >= budget.Period {
+		// Calculate how many periods have passed
+		periodsPassed := timeElapsed / budget.Period
+
+		if periodsPassed > 0 {
+			// Calculate the amount to distribute for all passed periods
+			coinsToDistribute := math.NewInt(int64(periodsPassed)).Mul(budget.TotalBudget.Amount.QuoRaw(int64(budget.Tranches)))
+			amount := sdk.NewCoin(budget.TotalBudget.Denom, coinsToDistribute)
+
+			// update the budget's remaining tranches
+			budget.Tranches -= periodsPassed
+
+			// update the TotalBudget amount
+			budget.TotalBudget.Amount = budget.TotalBudget.Amount.Sub(coinsToDistribute)
```

Comment:
totalbudget should probably stay the same but there be a claimed amount. Calculating the total budget based off the previous proposal would make it harder for clients. Since we are doing the right here as well, adding another uint64 doesnt seem like much overhead